### PR TITLE
Detect remotely closed sockets and add a single connection retry in the executor

### LINKED ISSUE #5538: Cached connections cause an unnecessary error when an idle session is reused after a connection failure
With long running, mostly idle connections, unexpected errors might show up to the user.

Given a scenario like the following:
 - user opens a session
 - user queries a distributed table
 - user is idle
 - a worker restarts
 - user invokes a query
 - user gets a connection error

Ideally the user would not see this error, as long as during their idle time there was no open transaction.

The underlying reason for this error is that cached connections are assumed to be in working condition. Unfortunately this assumption is not always true. Instead of assuming we should verify the liveness of the connection -to the best of our abilities- when we take a cached connection from the cache.

To reproduce this locally run the following: (commands are assuming you use citus_dev and are in the directory where you would normally run the citus_dev commands)

```bash
$ citus_dev make citus
$ psql -p9700
```

```sql
CREATE TABLE foo (a bigserial primary key, b text);
SELECT create_distributed_table('foo', 'a', shard_count => 4); -- just to limit some complexity during debugging we run with 4 shards
SELECT count(*) FROM foo;

-- keep this session open
```

```bash
# in a different shell, to simulate a restart
pg_ctl  restart -D sokeepalive/worker1 -o "-p 9702" -l worker1_logfile
```

```sql
-- back in initial psql session
SELECT count(*) FROM foo;
-- here you will observe a connection error, even though the worker has been fully restarted
-- expected behaviour should be a query that succeeded.
```

During investigation I noticed that the socket of the broken `pgconn struct` has data to be read. This has been observed via calling the internal libpq function `pqReadReady(connection->pgConn)`. As long as there is data to read this function returns a `1`, once all data is consumed, either a `0` or `-1` is returned. When an error is encountered on the socket a `-1` is returned. Also the result of `PQStatus(connection->pgCann)` changes from `CONNECTION_OK` to `CONNECTION_BAD`. Data can be consumed via `PQconsumeInput(connection->pgConn)`.

By consuming any data on a cached connection as long as there is data to be read we can then reliably check the status of the connection before returning it from cache.

### PR DESCRIPTION
**Pull request by onderkalaci**
PostgreSQL 15 exposes WL_SOCKET_CLOSED in WaitEventSet API, which is useful for detecting closed remote sockets. In this patch, we use this new event and try to detect closed remote sockets in the executor.

When a closed socket is detected, the executor now has the ability to retry the connection establishment. Note that, the executor can retry connection establishments only for the connection that has not been used. Basically, this patch is mostly useful for preventing the executor to fail if a cached connection is closed because of the worker node restart (or worker failover).

In other words, the executor cannot retry connection establishment if we are in a distributed transaction AND any command has been sent over the connection. That requires more sophisticated retry mechanisms. For now, fixing the above use case is enough.

Fixes #5538 

Earlier discussions: #5908, #6259 and #6283

### Summary of the current approach regards to earlier trials

As noted, we explored some alternatives before getting into this. https://github.com/citusdata/citus/pull/6283 is simple, but lacks an important property. We should be checking for `WL_SOCKET_CLOSED` _before_ sending anything over the wire. Otherwise, it becomes very tricky to understand which connection is actually safe to retry. For example, in the current patch, we can safely check `transaction->transactionState == REMOTE_TRANS_NOT_STARTED` before restarting a connection.

#6259 does what we intent here (e.g., check for sending any command). However, as @marcocitus noted, it is very tricky to handle `WaitEventSets` in multiple places. And, the executor is designed such that it reacts to the events. So, adding anything `pre-executor` seemed too ugly.

In the end, I converged into this patch. This patch relies on the simplicity of #6283 and also does a very limited handling of `WaitEventSets`, just for our purpose. Just before we add any connection to the execution, we check if the remote session has already closed. With that, we do a brief interaction of multiple wait event processing, but with different purposes. The new wait event processing we added does not even consider cancellations. We let that handled by the main event processing loop.

### CODE REVIEW

File: src/test/regress/sql/detect_conn_close.sql [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+SET citus.shard_count TO 8;
+SET citus.shard_replication_factor TO 1;
+
+CREATE TABLE socket_test_table(id bigserial, value text);
+SELECT create_distributed_table('socket_test_table', 'id');
+INSERT INTO socket_test_table (value) SELECT i::text FROM generate_series(0,100)i;
+
+-- first, simulate that we only have one cached connection per node
+SET citus.max_adaptive_executor_pool_size TO 1;
+SET citus.max_cached_conns_per_worker TO 1;
```
Comment (onderkalaci) [inline src/test/regress/sql/detect_conn_close.sql:24]: add test with max_cached_conns_per_worker=0

File: src/backend/distributed/executor/adaptive_executor.c [hunk 2]
```diff
@@ -5390,6 +5538,50 @@ BuildWaitEventSet(List *sessionList)
 }
 
 
+static void
+AddSessionToWaitEventSet(WorkerSession *session, WaitEventSet *waitEventSet)
```
Comment (onderkalaci) [inline src/backend/distributed/executor/adaptive_executor.c:5546]: copy & paste from above

File: src/backend/distributed/executor/adaptive_executor.c [hunk 3]
```diff
@@ -2450,6 +2472,74 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
...
 
 
+/*
+ * RemoteSocketClosedForNewSession is a helper function for detecting whether
+ * the remote socket corresponding to the input session is closed. This is
+ * mostly common there is a cached connection and remote server restarted
+ * (due to failover or restart etc.).
+ *
+ * The function is not a generic function that can be called at the start of
+ * the execution.
```
Comment (JelteF) [inline src/backend/distributed/executor/adaptive_executor.c:2482]: Why is it not a generic function?

File: src/backend/distributed/executor/adaptive_executor.c [hunk 4]
```diff
@@ -3647,13 +3775,38 @@ ConnectionStateMachine(WorkerSession *session)
 
 			case MULTI_CONNECTION_LOST:
 			{
-				/* managed to connect, but connection was lost */
-				workerPool->activeConnectionCount--;
```
Comment (JelteF) [inline src/backend/distributed/executor/adaptive_executor.c:3651]: Don't we we still need the old behavior in the else case?

Comment (JelteF) [reply src/backend/distributed/executor/adaptive_executor.c:3651]: Actually I now see this was moved to MULTI_CONNECTION_FAILED. Is there another way to reach that state?

File: src/backend/distributed/executor/adaptive_executor.c [hunk 5]
```diff
@@ -4164,7 +4326,13 @@ UpdateConnectionWaitFlags(WorkerSession *session, int waitFlags)
 		return;
 	}
 
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+	/* always detect closed sockets */
+	connection->waitFlags = waitFlags | WL_SOCKET_CLOSED;
```
Comment (JelteF) [inline src/backend/distributed/executor/adaptive_executor.c:4332]: Should the changes to this function be backported? i.e. by not waiting for WL_SOCKET_CLOSED do we miss failures?

File: src/test/regress/sql/detect_conn_close.sql [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+\if :server_version_ge_15
+\else
+\q
+\endif
+
+CREATE SCHEMA socket_close;
+SET search_path TO socket_close;
+
+SET citus.shard_count TO 8;
+SET citus.shard_replication_factor TO 1;
```
Comment (JelteF) [inline src/test/regress/sql/detect_conn_close.sql:16]: Missing citus.next_shard_id

In general I would recommend using `src/test/regress/bin/create_test.py` whenever you create a new test.

File: src/test/regress/expected/detect_conn_close.out [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+---------------------------------------------------------------------
+   115
+(1 row)
+
+-- kill all the cached backends on the workers
+SELECT result FROM run_command_on_workers($$SELECT count(*) FROM (SELECT pg_terminate_backend(pid, 1000) FROM pg_stat_activity WHERE query ilike '%socket_test_table%' AND pid !=pg_backend_pid()) as foo$$);
+ result
+---------------------------------------------------------------------
+ 0
+ 0
```
Comment (JelteF) [inline src/test/regress/expected/detect_conn_close.out:141]: It's not actually killing any connections here, probably the filter is incorrect

File: src/test/regress/expected/detect_conn_close.out [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+(1 row)
+
+-- also, recover insert .. select repartitioning
+INSERT INTO socket_test_table SELECT value::bigint, value FROM socket_test_table;
+-- kill all the cached backends on the workers
+SELECT result FROM run_command_on_workers($$SELECT count(*) FROM (SELECT pg_terminate_backend(pid, 1000) FROM pg_stat_activity WHERE query ilike '%socket_test_table%' AND pid !=pg_backend_pid()) as foo$$);
+ result
+---------------------------------------------------------------------
+ 0
+ 0
```
Comment (JelteF) [inline src/test/regress/expected/detect_conn_close.out:158]: Again no connections are actually killed by this.

File: src/test/regress/expected/detect_conn_close.out [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+---------------------------------------------------------------------
+     1
+(1 row)
+
+-- kill all the cached backends on the workers
+SELECT result FROM run_command_on_workers($$SELECT count(*) FROM (SELECT pg_terminate_backend(pid, 1000) FROM pg_stat_activity WHERE query ilike '%socket_test_table%' AND pid !=pg_backend_pid()) as foo$$);
+ result
+---------------------------------------------------------------------
+ 0
+ 0
```
Comment (JelteF) [inline src/test/regress/expected/detect_conn_close.out:175]: And again no connections are killed

**Review by JelteF — CHANGES_REQUESTED** [review]
Failure tests are failing due to giving a less useful error message.

File: src/test/regress/expected/detect_conn_close.out (hunk 7 above)
Comment (onderkalaci) [reply src/test/regress/expected/detect_conn_close.out:141]: ah, txes changes the query to COMMIT. Now, changed the filter to gpid, now all are terminated safely

I use boolean to make sure we do not introduce any flaky tests

File: src/backend/distributed/executor/adaptive_executor.c (hunk 3 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:2482]: There is nothing specific I'm concerned. But this function doesn't check for interrupts or does not wait at all. So, the callers should be careful not to expect those things

File: src/backend/distributed/executor/adaptive_executor.c (hunk 4 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:3651]: Hmm, that's a good point, we can directly come to the `MULTI_CONNECTION_FAILED` where we now decrement counter unnecessarily for such cases. Let me see what we can do there to fix that.

Though, practically speaking, Citus aggressively fails the execution of a connection fails, that's probably the reason no tests failed

File: src/backend/distributed/executor/adaptive_executor.c (hunk 5 above)
Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:4332]: I think we'd want these changes to go through a release testing given the points it touches

File: src/test/regress/sql/detect_conn_close.sql (hunk 6 above)
Comment (onderkalaci) [reply src/test/regress/sql/detect_conn_close.sql:31]: > Missing citus.next_shard_id

Should we still set  shardids even when we don't rely on them? I only set when explicitly needed

**Comment by onderkalaci** [comment]
> Failure tests are failing due to giving a less useful error message.

I might end-up storing the old error messages, otherwise there doesn't seem a way to give the same/similar errors.

If you have any other ideas, let me know

File: src/test/regress/sql/detect_conn_close.sql (hunk 6 above)
Comment (JelteF) [reply src/test/regress/sql/detect_conn_close.sql:31]: I guess it's fine not to set, but in my experience sometimes tests change later to rely on them.

File: src/backend/distributed/executor/adaptive_executor.c (hunk 3 above)
Comment (JelteF) [reply src/backend/distributed/executor/adaptive_executor.c:2482]: probably good to add those examples to the comment

**Comment by onderkalaci** [comment]
No performance impact is observed with this patch

**Pull request merged**